File: legacy/Sallie/backend/services/api-gateway/src/middleware/metrics.ts
import { Request, Response, NextFunction } from 'express';
import { register, Counter, Histogram, Registry } from 'prom-client';

// Create a registry for our metrics
const register = new Registry();

// Add default metrics (CPU, memory, etc.)
register.setDefaultLabels({
  app: 'sallie-api-gateway',
});

// Define custom metrics
const httpRequestsTotal = new Counter({
  name: 'http_requests_total',
  help: 'Total number of HTTP requests',
  labelNames: ['method', 'route', 'status_code'],
  registers: [register],
});

const httpRequestDuration = new Histogram({
  name: 'http_request_duration_seconds',
  help: 'Duration of HTTP requests in seconds',
  labelNames: ['method', 'route', 'status_code'],
  buckets: [0.1, 0.3, 0.5, 0.7, 1, 3, 5, 7, 10],
  registers: [register],
});

const activeConnections = new Counter({
  name: 'active_connections_total',
  help: 'Total number of active connections',
  registers: [register],
});

const serviceRequestsTotal = new Counter({
  name: 'service_requests_total',
  help: 'Total number of requests to backend services',
  labelNames: ['service', 'method', 'status_code'],
  registers: [register],
});

const serviceRequestDuration = new Histogram({
  name: 'service_request_duration_seconds',
  help: 'Duration of service requests in seconds',
  labelNames: ['service', 'method', 'status_code'],
  buckets: [0.1, 0.3, 0.5, 0.7, 1, 3, 5, 7, 10],
  registers: [register],
});

const authenticationTotal = new Counter({
  name: 'authentication_total',
  help: 'Total number of authentication attempts',
  labelNames: ['result'], // success, failure, invalid_token
  registers: [register],
});

const rateLimitHits = new Counter({
  name: 'rate_limit_hits_total',
  help: 'Total number of rate limit hits',
  labelNames: ['endpoint'],
  registers: [register],
});

// Middleware to track metrics
export const metricsMiddleware = (req: Request, res: Response, next: NextFunction) => {
  const start = Date.now();
  
  // Increment active connections
  activeConnections.inc();
  
  // Track response
  res.on('finish', () => {
    const duration = (Date.now() - start) / 1000;
    const route = req.route?.path || req.path;
    
    // Record HTTP request metrics
    httpRequestsTotal
      .labels(req.method, route, res.statusCode.toString())
      .inc();
    
    httpRequestDuration
      .labels(req.method, route, res.statusCode.toString())
      .observe(duration);
    
    // Decrement active connections
    activeConnections.dec();
  });
  
  next();
};

// Function to track service requests
export const trackServiceRequest = (service: string, method: string, statusCode: number, duration: number) => {
  serviceRequestsTotal
    .labels(service, method, statusCode.toString())
    .inc();
  
  serviceRequestDuration
    .labels(service, method, statusCode.toString())
    .observe(duration);
};

// Function to track authentication attempts
export const trackAuthentication = (result: 'success' | 'failure' | 'invalid_token') => {
  authenticationTotal.labels(result).inc();
};

// Function to track rate limit hits
export const trackRateLimit = (endpoint: string) => {
  rateLimitHits.labels(endpoint).inc();
};

// Function to get metrics for Prometheus
export const getMetrics = () => {
  return register.metrics();
};

// Function to reset all metrics (useful for testing)
export const resetMetrics = () => {
  register.clear();
};

export {
  register,
  httpRequestsTotal,
  httpRequestDuration,
  activeConnections,
  serviceRequestsTotal,
  serviceRequestDuration,
  authenticationTotal,
  rateLimitHits,
};

File: legacy/Sallie/backend/services/python-ai-service/src/middleware/metrics.py
from fastapi import Request, Response
from starlette.middleware.base import BaseHTTPMiddleware
from prometheus_client import Counter, Histogram, Gauge, generate_latest, CONTENT_TYPE_LATEST
import time
import logging
from ..core.config import settings

logger = logging.getLogger(__name__)

# Prometheus metrics
REQUEST_COUNT = Counter(
    'http_requests_total',
    'Total HTTP requests',
    ['method', 'endpoint', 'status_code']
)

REQUEST_DURATION = Histogram(
    'http_request_duration_seconds',
    'HTTP request duration in seconds',
    ['method', 'endpoint']
)

ACTIVE_CONNECTIONS = Gauge(
    'active_connections',
    'Number of active connections'
)

MODEL_REQUESTS = Counter(
    'model_requests_total',
    'Total model requests',
    ['model', 'request_type', 'status']
)

MODEL_RESPONSE_TIME = Histogram(
    'model_response_time_seconds',
    'Model response time in seconds',
    ['model', 'request_type']
)

class MetricsMiddleware(BaseHTTPMiddleware):
    """Middleware to collect Prometheus metrics"""
    
    async def dispatch(self, request: Request, call_next):
        start_time = time.time()
        
        # Increment active connections
        ACTIVE_CONNECTIONS.inc()
        
        try:
            response = await call_next(request)
            
            # Record request metrics
            method = request.method
            endpoint = request.url.path
            status_code = str(response.status_code)
            
            REQUEST_COUNT.labels(
                method=method,
                endpoint=endpoint,
                status_code=status_code
            ).inc()
            
            duration = time.time() - start_time
            REQUEST_DURATION.labels(
                method=method,
                endpoint=endpoint
            ).observe(duration)
            
            return response
            
        except Exception as e:
            # Record error metrics
            REQUEST_COUNT.labels(
                method=request.method,
                endpoint=request.url.path,
                status_code="500"
            ).inc()
            
            raise e
        finally:
            # Decrement active connections
            ACTIVE_CONNECTIONS.dec()
    
    @staticmethod
    def record_model_request(model: str, request_type: str, status: str, duration: float):
        """Record model-specific metrics"""
        MODEL_REQUESTS.labels(
            model=model,
            request_type=request_type,
            status=status
        ).inc()
        
        MODEL_RESPONSE_TIME.labels(
            model=model,
            request_type=request_type
        ).observe(duration)

# Function to get metrics endpoint
async def metrics_endpoint():
    """Prometheus metrics endpoint"""
    return Response(generate_latest(), media_type=CONTENT_TYPE_LATEST)

# Decorator for recording model metrics
def record_model_metrics(model: str, request_type: str):
    """Decorator to automatically record model metrics"""
    def decorator(func):
        async def wrapper(*args, **kwargs):
            start_time = time.time()
            status = "success"
            
            try:
                result = await func(*args, **kwargs)
                return result
            except Exception as e:
                status = "error"
                logger.error(f"Model request failed: {e}")
                raise
            finally:
                duration = time.time() - start_time
                MetricsMiddleware.record_model_request(model, request_type, status, duration)
        
        return wrapper
    return decorator
